### PR TITLE
Android: add cross-compilation configuration file flag

### PR DESCRIPTION
**Pull request by buttaface**
### Motivation:

This enables cross-compiling SPM for other hosts, such as Android.

### Modifications:

In addition to `--cross-compile-hosts`, add a new bootstrap flag, `--cross-compile-config`, that will pass in a JSON file with the destination cross-compilation flags.

### Result:

SPM can be cross-compiled for another OS, with Android as the first example.

### CODE REVIEW

File: Utilities/bootstrap
Comment (buttaface) [inline Utilities/bootstrap:151]: Even though these two are largely the same, they need to be formatted differently, so [I do that externally and centralize it in `build-script`](https://github.com/apple/swift/pull/33724/files#diff-cc09c8923370b808ebb7439cf4be9d2dde15e81796e80161576ae4cee0502ae7R155).

**Comment by tomerd** [comment]
cc @compnerd

File: Utilities/bootstrap
Comment (buttaface) [inline Utilities/bootstrap:313]: Rather than reuse this function to cross-compile the PackageDescription libraries, I extracted that portion into `build_packagedescription_libs_with_cmake()` above and this function is only used to natively compile SPM with CMake now.

**Comment by neonichu** [comment]
> Even though these two will be largely the same, both are needed because some PackageDescription libraries are not built by SPM itself, so the string is needed to build those with CMake.

I think this is actually a bug and I was planning to fix it. Any binaries we are installing should be built during the second stage, anything build by CMake should only be there temporarily.

**Comment by buttaface** [comment]
> I think this is actually a bug and I was planning to fix it.

OK, [I'm referring specifically to these libraries](https://github.com/apple/swift-package-manager/blob/6af862d3a7efb3108334be0d3e3e33641c91f267/Utilities/bootstrap#L397) that [are only built by CMake right now](https://github.com/apple/swift-package-manager/blob/6af862d3a7efb3108334be0d3e3e33641c91f267/Utilities/bootstrap#L413).

**Comment by neonichu** [comment]
Yep, we're talking about the same thing. I realize that they are currently being built by CMake only. We found this issue independently of this PR and I am planning to fix it soon, that's why I wanted to bring it up here.

**Comment by buttaface** [comment]
Yes, if two of those libraries were generated by the SPM build also (one of the PD libraries is built by SPM, but not installed), then the first flag added by this pull isn't needed. It is only needed because the build currently installs those CMake-generated libs. I can wait if you want to fix the build first.

File: Utilities/bootstrap
Comment (compnerd) [inline Utilities/bootstrap:206]: I don't think that this is correct; there are a number of targets that may be used, can you just invert the condition here instead?

```
if args.cross_compile_hosts:
  if re.match('android-', args.cross_compile_hosts):
    args.target_dir = os.path.join(args.build_dir, get_build_targets(args, cross_compile=True))
  else:
    args.target_dir = os.path.join(args.build_dir, 'apple/Products')
```

Yes, it isn't great, but leaves the behaviour the same while adding support for your use case.

**Review by compnerd — COMMENTED** [review]
Could you please also reflow the text that is being added?  Some of the lines are pretty long.

I'm not a big fan of adding more infrastructure in the bootstrap script, my understanding is that we want to move away from that.

File: Utilities/bootstrap
Comment (buttaface) [reply Utilities/bootstrap:206]: No, this is right, [see the similar logic at the end of the script](https://github.com/apple/swift-package-manager/blob/6af862d3a7efb3108334be0d3e3e33641c91f267/Utilities/bootstrap#L805). Basically, Mishal only added support for cross-compiling SPM for macOS arm64 and he explicitly spelled that out below, but didn't bother here: I went ahead and made it explicit here too, now that I added Android also.

**Comment by buttaface** [comment]
> Could you please also reflow the text that is being added?

Done, although this script already has a bunch of long lines that I didn't add.

> I'm not a big fan of adding more infrastructure in the bootstrap script

I'm not sure what "infrastructure" refers to in this context.

File: Utilities/bootstrap
Comment (compnerd) [reply Utilities/bootstrap:206]: WFM

**Review by compnerd — APPROVED** [review]

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by buttaface** [comment]
CI failures in the lldb tests on linux and building the compiler on macOS are unrelated to this pull.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@neonichu have your concerns here been addressed and this is ready to be merged?

**Comment by tomerd** [comment]
@buttaface please note we have branched 5.5 to release/5.5 so once this is merged we will need a corresponding PR if we need this to be included in 5.5

**Comment by buttaface** [comment]
Sigh, I made one last formatting change without testing and Python's complaining about that on the CI: I will never understand where line breaks are allowed in Python.

This pull works with the way the SPM build currently builds some release libraries with CMake and most everything else with the bootstrap SPM, but Boris wants to move the SPM build to a full SPM bootstrap. I'm fine with waiting for that change because it will allow me to simplify this pull and remove one of the bootstrap script flags I otherwise need, provided changing the build to SPM only doesn't take months or something.

I will submit a pull for the 5.5 branch once this is in.

**Comment by buttaface** [comment]
@abertelrud, I understand that you're looking to fix #3431 now by merging the two PD libraries, any plans to merge that back to the 5.5 branch? If not, I'd like to get this pull into trunk and 5.5 now and I will rework these changes, that cross-compile the two PD libraries with CMake, in another pull to adapt to when you get that PD merge pull into trunk.

If you plan to merge the two PD libraries in 5.5 too, then I will wait till you get that into trunk and rework this pull then.

**Comment by abertelrud** [comment]
I put up https://github.com/apple/swift-package-manager/pull/3464 now, though it's still a WIP.  I'm not sure whether it's in scope for 5.5 but I think it would be good if the question there can be cleared up (about unsafe compiler flags in the manifest).  I will wrap up the bootstrap script changes there today or tomorrow and then there should be a discussion about whether this can go into 5.5.

**Comment by abertelrud** [comment]
I don't have an opinion about whether this is merged first though — I think that the bootstrap script will need to be modified either way, and if this is merged first, I can easily adapt https://github.com/apple/swift-package-manager/pull/3464.

**Comment by buttaface** [comment]
Alright, I will wait till a decision is made on whether #3464 should go into 5.5 and then modify this as needed.

**Comment by abertelrud** [comment]
> Alright, I will wait till a decision is made on whether #3464 should go into 5.5 and then modify this as needed.

Sounds great.  I have completed my changes in the PR to unify the PackageDescription libraries, and I believe that that's ready to merge to mainline.  After that is approved and merged to main branch, then I'd suggest you rebase this PR on that and we'll try to get this merged.

I believe that the PackageDescription change should also go into 5.5 because it's needed in order to enable universal builds of PackageDescription and PackagePlugin.  So I'll be advocating strongly for that once it's been merged to main, and your follow-on 5.5 version of this PR could then build on that (just as in main).

File: Utilities/bootstrap
Comment (abertelrud) [inline Utilities/bootstrap:842]: I realize that this is existing code, but as I read it, this will only do the right thing if I'm building for Apple Silicon on an Intel machine.  Building a universal toolchain on an Apple Silicon machine doesn't seem to be supported, since `built_target` won't be `x86_64-apple-macosx`.  Am I misreading this?

Comment (abertelrud) [reply Utilities/bootstrap:842]: @shahmishal Do you know?

File: Utilities/bootstrap
Comment (abertelrud) [inline Utilities/bootstrap:628]: I think this is probably the only thing that needs to change after https://github.com/apple/swift-package-manager/pull/3464 is merged.

File: Utilities/bootstrap
Comment (buttaface) [reply Utilities/bootstrap:842]: Yeah, as I noted in a review comment for this pull earlier, only cross-compiling from mac x86_64 to mac arm64 is currently supported. It will error here if you go the other way.

File: Utilities/bootstrap
Comment (buttaface) [inline Utilities/bootstrap:481]: Now that the only use of this `targets` parameter was removed below, as the PD libraries no longer need to be built by CMake, I removed it altogether.

**Comment by buttaface** [comment]
Updated this pull to adapt to #3431, it will need to be merged after that pull.

**Review by headdoctor77 — APPROVED** [review]

**Comment by buttaface** [comment]
Now that #3431 is in, I tested this with the latest trunk source snapshot and confirmed that everything cross-compiles, with only the `.swiftinterface` files now missing as in that pull.

I think this is ready to go in.

**Comment by buttaface** [comment]
Ping, can we get this in? I'd like to get this into the 5.5 branch too.

**Comment by abertelrud** [comment]
> Now that #3431 is in, I tested this with the latest trunk source snapshot and confirmed that everything cross-compiles, with only the `.swiftinterface` files now missing as in that pull.

FWIW, https://github.com/apple/swift-package-manager/pull/3528 adds back the `.swiftinterface` files but it still needs to be reviewed.

**Comment by abertelrud** [comment]
> Ping, can we get this in? I'd like to get this into the 5.5 branch too.

I will put up PRs to ask to merge https://github.com/apple/swift-package-manager/pull/3464 and https://github.com/apple/swift-package-manager/pull/3528 into 5.5.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
> > Ping, can we get this in? I'd like to get this into the 5.5 branch too.
> 
> I will put up PRs to ask to merge #3464 and #3528 into 5.5.

I have put up https://github.com/apple/swift-package-manager/pull/3536 with cherry-picks for this.

**Comment by buttaface** [comment]
@abertelrud, mind merging this? I will submit for the 5.5 branch once this in trunk.

**Review by abertelrud — APPROVED** [review]

**Comment by abertelrud** [comment]
This looks fine, but do you think there is a problem that I had to temporarily deactivate the changes to build universal PackageDescription and PackagePlugin (https://github.com/apple/swift-package-manager/pull/3540)?  I don't think it will impact what you're doing here, but what do you think?

**Comment by buttaface** [comment]
Yeah, I saw that: it will affect this pull as I'm removing the CMake cross-compilation of the PD/PP libraries for non-macOS platforms here, whereas you went back to installing those CMake-built libraries there. But that's fine, I'd rather just get this in and I figure you'll switch that back to the SPM build later.

**Comment by buttaface** [comment]
Ping, would be good to get this in.

**Comment by abertelrud** [comment]
Looks good to me given the current somewhat unfortunate implementation of universal binaries for the macOS toolchain.  I have requested that the compiler scripts change to using `--build-all-architectures` to convey the intent of building for all supported architectures on platforms where that is relevant, and that the `--cross` flags be used for building for other platforms, as is done here.

@tomerd @neonichu any concerns with this change to unblock Android?

**Comment by abertelrud** [comment]
Actually:  @buttaface what is the semantic distinction here between `cross-compile-hosts` and `cross-compile-target-triple`?  It sounds as if the latter is essentially the way the former was intended to work, except that it only supports a single target.  Is the reason for keeping `cross-compile-hosts` here because there are existing scripts that call it?

**Comment by buttaface** [comment]
Ideally, nothing different. `--cross-compile-hosts` uses the `build-script` format of `android-aarch64`, while `--cross-compile-target-triple` is the LLVM/Swift target triple, ie `aarch64-unknown-linux-android`. We could combine the two to only use the latter, but since the former is already being used as you noted, I just added the triple as another flag.

**Comment by abertelrud** [comment]
> Ideally, nothing different. `--cross-compile-hosts` uses the `build-script` format of `android-aarch64`, while `--cross-compile-target-triple` is the LLVM/Swift target triple, ie `aarch64-unknown-linux-android`. We could combine the two to only use the latter, but since the former is already being used as you noted, I just added the triple as another flag.

Thanks for clarifying.  I think it makes sense to merge this to unblock Android, but over time it seems that the script can be cleaned up.  I think that can be done in separate PRs.  Would it make sense to make `--cross-compile-target-triple` be able to support a list so that it completely subsume `--cross-compile-hosts`, or am I missing the point of that new flag?  i.e. if it were a list (which AFAICT should be a simple change), could we get the toolchain build scripts to use it and then get rid of `--cross-compile-hosts`?

**Comment by buttaface** [comment]
[I actually don't think `--cross-compile-hosts` should be a list](https://github.com/apple/swift/pull/33724/#discussion_r533554556), but yes, we could probably merge these two flags at some point. I have opened a pull in the main Swift repo to pass these cross-compilation flags into SPM, apple/swift#36917, I will raise the possibility with Mishal there.

**Comment by buttaface** [comment]
Hold on, I just thought of a way to do this without that additional flag, will try it and update this pull if it works.

**Comment by abertelrud** [comment]
> Hold on, I just thought of a way to do this without that additional flag, will try it and update this pull if it works.

That would be great!  I don't want to hold this up any longer, but any simplification to what's already a complicated situation would be great.

**Comment by abertelrud** [comment]
> [I actually don't think `--cross-compile-hosts` should be a list](https://github.com/apple/swift/pull/33724/#discussion_r533554556), . . .

Thanks for the reference to that discussion.  I think this gets to a concern I have with the terminology here, and the conflation of platform and architecture when speaking of cross-compilation.  In my view it doesn't make sense to cross-compile for, say, Windows and Darwin or Android and Linux in a single invocation of the bootstrap script, so really, only one target platform makes sense.

But within that platform, it makes sense to either choose the architectures for which to compile or have a "default vs all" switch.  While the target triples are currently presented as a flat list, it's really only a single platform with one or more architectures that makes any sense for a single given build.

I think ideally it should be possible to specify a target platform and then either a set of target architectures for that platform, or perhaps simpler, a single flag that says "build all architectures for the platform" as opposed to building just a single default architecture (which would, for example, be the architecture of the host).  Perhaps that won't suffice for some platforms, where it isn't the host, and there are also multiple architectures to choose from.

File: Utilities/bootstrap
Comment (buttaface) [inline Utilities/bootstrap:290]: Alright, this works and allows me to remove the extra triple flag. I just realized that the destination.json is just a JSON file, so we can parse it to get the target triple.

@abertelrud, if you agree with this change, just let me know and I will squash it into the first commit and update the commit message and we can get this in.

**Comment by buttaface** [comment]
> I think ideally it should be possible to specify a target platform and then either a set of target architectures for that platform

Yeah, we could reorganize it that way. As I mentioned before, I think `--cross-compile-hosts` was a quick hack by Boris to get the toolchain building for Apple Silicon, we can upgrade it with time.

File: Utilities/bootstrap
Comment (abertelrud) [reply Utilities/bootstrap:290]: Thanks @buttaface, simplifying this down to get the information from the JSON instead of the having to pass another flag is a good improvement.  I still want to hear what @neonichu and @tomerd think but this looks good given how things currently work.  I will still want to work with @shahmishal to see if we can simplify the conceptual model around building for various platforms and architectures.

**Comment by abertelrud** [comment]
> > I think ideally it should be possible to specify a target platform and then either a set of target architectures for that platform
> 
> Yeah, we could reorganize it that way. As I mentioned before, I think `--cross-compile-hosts` was a quick hack by Boris to get the toolchain building for Apple Silicon, we can upgrade it with time.

Absolutely, it was all very understandable at the time.  Just thoughts about how to improve on this step-by-step.

File: Utilities/bootstrap
Comment (buttaface) [reply Utilities/bootstrap:290]: Alright, will squash this commit and update the message.

**Comment by buttaface** [comment]
Ready to go.

**Comment by buttaface** [comment]
@neonichu, I think you just need to sign off on this? I have updated my build-script pull, apple/swift#36917, that shows how this new bootstrap flag will be invoked.

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@neonichu @buttaface This is currently marked with the `wip` (work-in-progress) tag.  Is it ready to merge?

**Comment by buttaface** [comment]
Yeah, that tag was added two months ago, when this pull wasn't building. Several changes were made since then, please remove the tag and merge.

**Comment by neonichu** [comment]
Thanks for the PR @buttaface!

**Comment by buttaface** [comment]
Thanks for finally getting this in, I will submit for the 5.5 branch.

**Pull request merged**
